src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,377 @@
+                return true;
+            }
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '&'))
+            {
+                rankOrModifier = ByRef;
+                return true;
+            }
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '['))
+            {
+                // SZArray := []
+                // MDArray := [*] or [,] or [,,,, ...]
+
+                int rank = 1;
+                bool hasSeenAsterisk = false;
+
+            ReadNextArrayToken:
+
+                if (TryStripFirstCharAndTrailingSpaces(ref input, ']'))
+                {
+                    // End of array marker
+                    rankOrModifier = rank == 1 && !hasSeenAsterisk ? SZArray : rank;
+                    return true;
+                }
+
+                if (!hasSeenAsterisk)
+                {
+                    if (rank == 1 && TryStripFirstCharAndTrailingSpaces(ref input, '*'))
+                    {
+                        // [*]
+                        hasSeenAsterisk = true;
+                        goto ReadNextArrayToken;
+                    }
+                    else if (TryStripFirstCharAndTrailingSpaces(ref input, ','))
+                    {
+                        // [,,, ...]
+                        checked { rank++; }
+                        goto ReadNextArrayToken;
+                    }
+                }
+
+                // Don't know what this token is.
+                // Fall through to 'return false' statement.
+            }
+
+            input = originalInput; // ensure 'ref input' not mutated
+            rankOrModifier = 0;
+            return false;
+        }
+
+        internal static bool TryStripFirstCharAndTrailingSpaces(ref ReadOnlySpan<char> span, char value)
+        {
+            if (!span.IsEmpty && span[0] == value)
+            {
+                span = span.Slice(1).TrimStart();
+                return true;
+            }
+            return false;
+        }
+
+        internal static InvalidOperationException InvalidOperation_MaxNodesExceeded(int limit) =>
+#if SYSTEM_REFLECTION_METADATA || SYSTEM_PRIVATE_CORELIB
+            new InvalidOperationException(SR.Format(SR.InvalidOperation_MaxNodesExceeded, limit));
+#else // tools that reference this file as a link
+            new InvalidOperationException();
+#endif
+
+        internal static ArgumentException ArgumentException_InvalidTypeName(int errorIndex) =>
+#if SYSTEM_PRIVATE_CORELIB
+            new ArgumentException(SR.Arg_ArgumentException, $"typeName@{errorIndex}");
+#elif SYSTEM_REFLECTION_METADATA
+            new ArgumentException(SR.Argument_InvalidTypeName);
+#else // tools that reference this file as a link
+            new ArgumentException();
+#endif
+
+        internal static InvalidOperationException InvalidOperation_NotGenericType() =>
+#if SYSTEM_REFLECTION_METADATA || SYSTEM_PRIVATE_CORELIB
+            new InvalidOperationException(SR.InvalidOperation_NotGenericType);
```

Comment:
The internal uses of TypeName in CoreLib and tools should never hit these InvalidOperationException conditions. We can make them assert and/or throw exception without a message. We do not need to pay for these unreachable messaged in CoreLib given that this is under ifdef anyway.

Reply:
> We do not need to pay for these unreachable messaged in CoreLib given that this is under ifdef anyway.

I've reused the resources used by the reflection APIs:

https://github.com/dotnet/runtime/blob/33e3f0c746c38e10b701932d14230830647dddc6/src/libraries/System.Private.CoreLib/src/Resources/Strings.resx#L2681-L2683

> and tools should never hit these

As soon as tools update to newer SDK I am going to remove a bunch of `#ifdefs` and rely on public APIs only.